### PR TITLE
Misc. bug fixes and enhancements to subrobot machinery

### PR DESCRIPTION
**Pull request by mass2010chromium**
See https://github.com/krishauser/Klampt/pull/141#pullrequestreview-940107279

### CODE REVIEW

**Comment by mass2010chromium** [comment]
maybe don't merge this yet because there will probably be more patches as i start testing new motion code with klampt

**Comment by mass2010chromium** [comment]
Found bug (or at least weird behavior) with IKSolver and subrobots -- IKSolver destroys subrobot information; created a new class that wraps IKSolver but does the config transforming correctly

File: Python/klampt/model/ik.py [hunk 1]
```diff
@@ -289,6 +289,122 @@ def objects(objectives):
     raise NotImplementedError()
     pass
 
+class SubRobotIKSolver(IKSolver):
```
Comment (mass2010chromium) [inline Python/klampt/model/ik.py:292]: Subclass of IKSolver that keeps metadata important to subrobots. Wraps IKSolver functions to expose only subrobot dofs, and allow setting of only subrobot dofs.

File: Python/klampt/model/ik.py [hunk 2]
```diff
@@ -353,8 +469,7 @@ def solver(
             res = []
             for key,(r,objs) in robs.items():
                 if isinstance(r,SubRobotModel):
-                    s = IKSolver(r._robot)
-                    s.setActiveDofs(r._links)
+                    s = SubRobotIKSolver(r)
```
Comment (mass2010chromium) [inline Python/klampt/model/ik.py:472]: Replace IKSolver with SubRobotIKSolver whenever ik solve is requested using a subrobot.

File: Python/klampt/model/subrobot.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def __init__(self, robot : Union[RobotModel,'SubRobotModel'], links : Sequence[U
         """
         assert isinstance(robot,(RobotModel,SubRobotModel)),"SubRobotModel constructor must be given a RobotModel or SubRobotModel as first argument"
         self._robot = robot
-        self._links = links[:]     #type : List[int]
+        self._links = [ robot.link(i).index for i in links ]     #type : List[int]
```
Comment (mass2010chromium) [inline Python/klampt/model/subrobot.py:36]: Allow passing input links as strings/link names (as is specified in the docstring). Internally, the representation is always link indices.

File: Python/klampt/model/ik.py (hunk 1 above)
Comment (mass2010chromium) [inline Python/klampt/model/ik.py:292]: Subclass of IKSolver that keeps metadata important to subrobots. Wraps IKSolver functions to expose only subrobot dofs, and allow setting of only subrobot dofs.

File: Python/klampt/model/ik.py (hunk 2 above)
Comment (mass2010chromium) [inline Python/klampt/model/ik.py:472]: Replace IKSolver with SubRobotIKSolver when subrobots are passed to IK solve routines.

File: Python/klampt/model/subrobot.py (hunk 3 above)
Comment (mass2010chromium) [inline Python/klampt/model/subrobot.py:36]: Allow passing of link names or link indices (as specified in docstring). Internally, link indices are used.

File: Python/klampt/model/subrobot.py [hunk 4]
```diff
@@ -46,15 +46,16 @@ def __init__(self, robot : Union[RobotModel,'SubRobotModel'], links : Sequence[U
         for i,l in enumerate(self._links):
             if isinstance(l,str):
                 self._links[i] = robot.link(l).getIndex()
+
         self._inv_links = dict((l,i) for (i,l) in enumerate(self._links))
 
-    def tofull(self,object,reference=None):
+    def tofull(self,obj,reference=None):
```
Comment (mass2010chromium) [inline Python/klampt/model/subrobot.py:52]: Fix style: `object` is a python keyword.

File: Python/klampt/model/subrobot.py [hunk 5]
```diff
@@ -63,44 +64,45 @@ def tofull(self,object,reference=None):
...
+            if isinstance(obj,Trajectory):
+                if isinstance(obj,HermiteTrajectory):
                     raise NotImplementedError("Can't lift hermite trajectories to full robots yet")
-                newmilestones = self.tofull(object.milestones,reference=reference)
-                return object.constructor()(object.times,newmilestones)
+                newmilestones = self.tofull(obj.milestones,reference=reference)
+                return obj.constructor()(obj.times,newmilestones)
             else:
-                raise ValueError("Invalid object type, not an integer, configuration, or Trajectory")
+                raise ValueError(f"Invalid obj type ({obj}, type {type(obj)}), not an integer, configuration, or Trajectory")
```
Comment (mass2010chromium) [inline Python/klampt/model/subrobot.py:105]: More informative error message, specifying the types in the type error.

File: Python/klampt/model/subrobot.py [hunk 6]
```diff
@@ -159,6 +161,7 @@ def _computeDrivers(self):
             for l in d.getAffectedLinks():
                 if l in self._links:
                     self._drivers.append(d)
+                break
```
Comment (mass2010chromium) [inline Python/klampt/model/subrobot.py:164]: When computing drivers, don't duplicate drivers by the number of links they affect.

NOTE: unsure if this is intended behavior or not, but it makes things harder to work with.

**Review by mass2010chromium — COMMENTED** [review]
Summary of changes

**Comment by krishauser** [comment]
I have some changes, could you please put these into your PR
[pull_request.zip](https://github.com/krishauser/Klampt/files/8606619/pull_request.zip)
?

**Comment by mass2010chromium** [comment]
> I have some changes, could you please put these into your PR [pull_request.zip](https://github.com/krishauser/Klampt/files/8606619/pull_request.zip) ?

It looks like the `subrobot.py` in this zipfile is identical to `ik.py`; is this intended?

**Comment by krishauser** [comment]
Merged (manually) to master

**Pull request closed without merging**